### DISCUSSION TITLE
in yaml rules, is it possible to count nodes?

### DISCUSSION THREAD

**Discussion by emmanueltouzery** [ast-grep/ast-grep#1201] [Q&A] [answered]
in java, i'm interested in filtering by the number of child nodes of a certain type. concretely, i'm searching for methods with a certain number of parameters.

so far i was using more yaml rules and not really patterns. it seems this should be doable through patterns (`function($A, $B)`), but I can't make it work, but also i already have rules and I'd prefer to slightly extend them rather than adding patterns. If I have no choice of course..

Here is a playground link with an attempt at a pattern which doesn't work:
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiamF2YSIsInF1ZXJ5IjoiY2xhc3MgJEMge1xuICAgICQkJE1PRElGSUVSUyB0ZXN0RnVuY3Rpb24oJCRQLCAkJFEsICQkUikgeyAkJCQgfVxufVxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6IiIsInNvdXJjZSI6InB1YmxpYyBjbGFzcyBUZXN0IHtcbiAgcHVibGljIHN0YXRpYyB2b2lkIHRlc3RGdW5jdGlvbihTdHJpbmdbXSBjb21tYW5kcywgbG9uZyB0aW1lb3V0LCBUaW1lVW5pdCB0aW1lVW5pdCkge1xuICAgIHRlc3RGdW5jdGlvbihjb21tYW5kcywgdGltZW91dCwgdGltZVVuaXQsIExvZ0V4ZWN1dGlvbi5ZZXMpO1xuICB9XG5cbiAgcHVibGljIHN0YXRpYyBzeW5jaHJvbml6ZWQgdm9pZCB0ZXN0RnVuY3Rpb24oU3RyaW5nW10gY29tbWFuZHMsIGxvbmcgdGltZW91dCwgVGltZVVuaXQgdGltZVVuaXQsIExvZ0V4ZWN1dGlvbiBsb2dFeGVjdXRpb24pIHtcbiAgICAgIHJldHVybjtcbiAgfVxufSJ9

(the search should match only the first overload)

thank you!

**Comment by HerringtonDarkholme** [comment]
not directly, counting number is tracked in https://github.com/ast-grep/ast-grep/issues/676

**Reply by HerringtonDarkholme** [reply]
Alternatively, you can use `precedes` or `follows` rule to express the order of AST node.

**Reply by emmanueltouzery** [reply]
ah yes, i could generate an ad-hoc query everytime with the correct number of "precedes" rules.

that'll probably work! I'll test it and mark as answer if i make it work.

thank you!

**Reply by emmanueltouzery** [reply]
I feel that this should work, but it doesn't: [playground link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmEiLCJxdWVyeSI6IiIsInJld3JpdGUiOiIiLCJjb25maWciOiJ1dGlsczpcbiAgICBoYXMtdGhyZWUtcGFyYW1ldGVyczpcbiAgICAgICAgaW5zaWRlOlxuICAgICAgICAgICAga2luZDogbWV0aG9kX2RlY2xhcmF0aW9uXG4gICAgICAgIG1hdGNoZXM6IG1ldGhvZC1oYXMtdGhyZWUtcGFyYW1ldGVyc1xuICAgIFxuICAgIG1ldGhvZC1oYXMtdGhyZWUtcGFyYW1ldGVyczpcbiAgICAgICAgaGFzOlxuICAgICAgICAgICAgc3RvcEJ5OlxuICAgICAgICAgICAgICAgIGtpbmQ6IGZvcm1hbF9wYXJhbWV0ZXJcbiAgICAgICAgICAgIHByZWNlZGVzOlxuICAgICAgICAgICAgICAgIG1hdGNoZXM6IGlzLXNlY29uZC1wYXJhbWV0ZXJcbiAgICAgICAgICAgIG5vdDpcbiAgICAgICAgICAgICAgICBmb2xsb3dzOlxuICAgICAgICAgICAgICAgICAgICBraW5kOiBmb3JtYWxfcGFyYW1ldGVyXG5cbiAgICBpcy1zZWNvbmQtcGFyYW1ldGVyOlxuICAgICAgICBraW5kOiBmb3JtYWxfcGFyYW1ldGVyXG4gICAgICAgIHByZWNlZGVzOlxuICAgICAgICAgICAgbWF0Y2hlczogaXMtdGhpcmQtbGFzdC1wYXJhbWV0ZXJcblxuICAgIGlzLXRoaXJkLWxhc3QtcGFyYW1ldGVyOlxuICAgICAgICBraW5kOiBmb3JtYWxfcGFyYW1ldGVyXG4gICAgICAgIG5vdDpcbiAgICAgICAgICAgIHByZWNlZGVzOlxuICAgICAgICAgICAgICAgIGtpbmQ6IGZvcm1hbF9wYXJhbWV0ZXJcblxucnVsZTpcbiAgICBwYXR0ZXJuOiB0ZXN0RnVuY3Rpb25cbiAgICBhbGw6XG4gICAgICAgIC0gbWF0Y2hlczogaGFzLXRocmVlLXBhcmFtZXRlcnNcbiAgICBpbnNpZGU6XG4gICAgICAgIHN0b3BCeTpcbiAgICAgICAgICAgIGtpbmQ6IGNsYXNzX2RlY2xhcmF0aW9uXG4gICAgICAgIGhhczpcbiAgICAgICAgICAgIHBhdHRlcm46IFRlc3QiLCJzb3VyY2UiOiJwdWJsaWMgY2xhc3MgVGVzdCB7XG4gIHB1YmxpYyBzdGF0aWMgdm9pZCB0ZXN0RnVuY3Rpb24oU3RyaW5nW10gY29tbWFuZHMsIGxvbmcgdGltZW91dCwgVGltZVVuaXQgdGltZVVuaXQpIHtcbiAgICB0ZXN0RnVuY3Rpb24oY29tbWFuZHMsIHRpbWVvdXQsIHRpbWVVbml0LCBMb2dFeGVjdXRpb24uWWVzKTtcbiAgfVxuXG4gIHB1YmxpYyBzdGF0aWMgc3luY2hyb25pemVkIHZvaWQgdGVzdEZ1bmN0aW9uKFN0cmluZ1tdIGNvbW1hbmRzLCBsb25nIHRpbWVvdXQsIFRpbWVVbml0IHRpbWVVbml0LCBMb2dFeGVjdXRpb24gbG9nRXhlY3V0aW9uKSB7XG4gICAgICByZXR1cm47XG4gIH1cbn0ifQ==).

it is quite hairy... i start from the identifier of the method name. then i go to the parent, the method definition, and from the method definition search for the formal parameters, then try to establish there are exactly three. but it fails even finding the formal parameters. i'll return to this tomorrow. it seems workable but i'm lacking ast-grep experience on this (i feel) quite advanced case.

**Reply by HerringtonDarkholme** [reply, marked as answer]
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmEiLCJxdWVyeSI6IiIsInJld3JpdGUiOiIiLCJjb25maWciOiJ1dGlsczpcbiAgICBoYXMtdGhyZWUtcGFyYW1ldGVyczpcbiAgICAgICAgaW5zaWRlOlxuICAgICAgICAgICAga2luZDogbWV0aG9kX2RlY2xhcmF0aW9uXG4gICAgICAgIHByZWNlZGVzOlxuICAgICAgICAgICAgbWF0Y2hlczogdGhyZWUtcGFyYW1ldGVyc1xuICAgIFxuICAgIHRocmVlLXBhcmFtZXRlcnM6XG4gICAgICAgIGhhczpcbiAgICAgICAgICAgIGtpbmQ6IGZvcm1hbF9wYXJhbWV0ZXJcbiAgICAgICAgICAgIHN0b3BCeTpcbiAgICAgICAgICAgICAgICBraW5kOiBmb3JtYWxfcGFyYW1ldGVyXG4gICAgICAgICAgICBmb2xsb3dzOlxuICAgICAgICAgICAgICAgIG1hdGNoZXM6IGlzLXNlY29uZC1wYXJhbWV0ZXJcbiAgICAgICAgICAgICAgICBzdG9wQnk6XG4gICAgICAgICAgICAgICAgICAgIGtpbmQ6IGZvcm1hbF9wYXJhbWV0ZXJcbiAgICAgICAgICAgIHByZWNlZGVzOlxuICAgICAgICAgICAgICAgIHJlZ2V4OiAnXFwpJ1xuXG4gICAgaXMtc2Vjb25kLXBhcmFtZXRlcjpcbiAgICAgICAga2luZDogZm9ybWFsX3BhcmFtZXRlclxuICAgICAgICBmb2xsb3dzOlxuICAgICAgICAgICAgbWF0Y2hlczogaXMtZmlyc3RcbiAgICAgICAgICAgIHN0b3BCeTogXG4gICAgICAgICAgICAgICAga2luZDogZm9ybWFsX3BhcmFtZXRlclxuICAgIGlzLWZpcnN0OlxuICAgICAgICBraW5kOiBmb3JtYWxfcGFyYW1ldGVyXG4gICAgICAgIGZvbGxvd3M6IFxuICAgICAgICAgICAgcmVnZXg6ICdcXCgnXG4gICAgICAgICAgICAgICAgXG5ydWxlOlxuICAgIHBhdHRlcm46IHRlc3RGdW5jdGlvblxuICAgIGFsbDpcbiAgICAgICAgLSBtYXRjaGVzOiBoYXMtdGhyZWUtcGFyYW1ldGVyc1xuICAgIGluc2lkZTpcbiAgICAgICAgc3RvcEJ5OlxuICAgICAgICAgICAga2luZDogY2xhc3NfZGVjbGFyYXRpb25cbiAgICAgICAgaGFzOlxuICAgICAgICAgICAgcGF0dGVybjogVGVzdCIsInNvdXJjZSI6InB1YmxpYyBjbGFzcyBUZXN0IHtcbiAgcHVibGljIHN0YXRpYyB2b2lkIHRlc3RGdW5jdGlvbihTdHJpbmdbXSBjb21tYW5kcywgbG9uZyB0aW1lb3V0LCBUaW1lVW5pdCB0aW1lVW5pdCkge1xuICAgIHRlc3RGdW5jdGlvbihjb21tYW5kcywgdGltZW91dCwgdGltZVVuaXQsIExvZ0V4ZWN1dGlvbi5ZZXMpO1xuICB9XG5cbiAgcHVibGljIHN0YXRpYyBzeW5jaHJvbml6ZWQgdm9pZCB0ZXN0RnVuY3Rpb24oU3RyaW5nW10gY29tbWFuZHMsIGxvbmcgdGltZW91dCwgVGltZVVuaXQgdGltZVVuaXQsIExvZ0V4ZWN1dGlvbiBsb2dFeGVjdXRpb24pIHtcbiAgICAgIHJldHVybjtcbiAgfVxufSJ9

**Reply by emmanueltouzery** [reply]
extremely impressive, thank you very much!!

**Reply by HerringtonDarkholme** [reply]
On the main branch, `nthChild` has been implemented! This can greatly simplify the rule to this

```yml
id: find-overload
language: java
rule:
  pattern: testFunction
  precedes:
    has: { nthChild: 3 }
    not:
      has: { nthChild: 4 }
  inside:
    stopBy:
      kind: class_declaration
    has:
      pattern: Test
```

output:

<img width="1032" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/f58e6f00-2e28-4378-883f-a1f5589abbe3">

**Reply by emmanueltouzery** [reply]
Fantastic! I think I'll depend on this, I hadn't implemented the support yet. 

ast-grep really enables use cases that just weren't on the table before. Thank you!

**Reply by HerringtonDarkholme** [reply]
the nthChild rule is shipped in 0.23.0
